### DISCUSSION TITLE
Making it easier to enable debug logging

### DISCUSSION THREAD

**Discussion by bdraco** [home-assistant/architecture#715] [General] [answered]
In github issues we spend a lot of time explaining how to enable debug logging since it can take a few tries for users to get the logger name correct.  Sometimes issues go unsolved for the simple reason that users cannot clear this hurdle.

Proposal is to adding the following:

- A list of loggers to the manifest
- An option for the integration config entry to enable/disable debug logging

When the option is clicked, it will turn on debug logging for the loggers listed in the manifest as well as the integration itself (aka `homeassistant.components.DOMAIN`.  If debug logging is enabled for the integration (aka `homeassistant.components.DOMAIN`) the option will appear to disable debug logging and it will disable debug logging for the loggers and the integration itself.

Cavets:

If multiple integrations share the same loggers, and debug logging is enabled for multiple integrations, disabling debug logging could disable logging unexpectedly. This seems unlikely to be a problem in production, but we could solve it by not adding an option to disable debug logging, and only allowing them to enable it via the integration config entry. Disabling would have to be done by calling the service in this case.

**Comment by balloob** [comment]
Great idea.

ZHA already has this functionality. They use this to stream debug logs to the frontend while pairing.

If we adopt their strategy, we could stream the debug logs to the UI and show them, and have a download button at the bottom ([something like this](https://github.com/esphome/esp-web-tools/pull/165)). Debug logging would automatically disable when dialog is closed/websocket connection closes.

**Reply by dshokouhi** [reply]
Whats nice about this idea is that the android app can probably also forward logs back to core for users to debug issues related to the mobile app :) We have a logging section already in the app for on device logs but it may be more useful to a user to see things from both ends at once. Like sending a notification to the app and seeing it get processed or fail etc...

**Comment by thecode** [comment]
@bdraco I think it would be good to have an option to disable logs, because the same user who enabled them via the UI would not necessarily know how to disable them, to solve the the problem you raised there could be one **"Disable debug logs"** button, maybe placed at the server section near the restart button so that we can instruct users to go there to disable.
This will have fixed functionality and clear definition that clicking this button disable all manifest enabled logs and put the system back to what is defined in `configuration.yaml` (or the defaults).

**Comment by shbatm** [comment]
> An option for the integration config entry to enable/disable debug logging

Instead of just a tick box to enable `DEBUG` could it support setting different levels (`INFO`/`WARN`)? This may add a layer of complexity for a novice user to get the right setting, but would shift more of the logger yaml config to the UI and allow finer control without having to restart HA to affect the new settings.

**Reply by thecode** [reply]
> > An option for the integration config entry to enable/disable debug logging
> 
> Instead of just a tick box to enable `DEBUG` could it support setting different levels (`INFO`/`WARN`)? This may add a layer of complexity for a novice user to get the right setting, but would shift more of the logger yaml config to the UI and allow finer control without having to restart HA to affect the new settings.

We had a short discussion about it on Discord, I was also thinking we will need it, but during the discussion I realized that most of the time you would need debug level and the occasions when you don't are rare so in the worst case you just end up with bigger log, but unlike in `configuration.yaml` that the change is kinda fixed, here you can enable debug, get the logs and disable very easy.

**Reply by shbatm** [reply]
> We had a short discussion about it on Discord, I was also thinking we will need it, but during the discussion I realized that most of the time you would need debug level and the occasions when you don't are rare so in the worst case you just end up with bigger log, but unlike in `configuration.yaml`

I would agree with that. I was actually thinking more of the times when you want to shut-up a noisy component by switching it to a higher level (e.g HA is at INFO and you want some at WARN or ERROR only). It's easy enough to make the change in my YAML and do the restart, it would be more of a nice-to-have to be able to customize that live from the UI rather than waiting for a restart.

**Reply by thecode** [reply]
![image](https://user-images.githubusercontent.com/1858925/151075743-3528002c-e2e5-4f68-8429-94194f8a0f27.png)

**Reply by shbatm** [reply]
Thanks. I honestly totally forgot about the service...

**Comment by postlund** [comment]
This would be an *awesome* feature and I've been thinking about adding something similar for a while. A truly great addition would be forwarding logs to a separate file that can be downloaded afterwards. Basically press a "Record" button for an integration to start, press "End" when done and get a download button. As the main log file can be rather big and might also contain log entries for other things depending on user configuration, this would get right or log of noise.

**Comment by eifinger** [comment]
Would it also be possible to add other parts of homeassistant e.g. helper functions to the logger list?
I just fixed an issue for the `google_travel_time` and `waze_travel_time` integrations in regards to how they resolve the actual address/location to use.
https://github.com/home-assistant/core/pull/66081

If people enable debug logging for waze_travel_time it would be great if debug logs from `homeassistant/helpers/location` could be included.

**Reply by thecode** [reply]
How did you enable the debug logs for `homeassistant/helpers/location`?, you can add the component you need to the loggers and it will also be set to `debug` level when this is implemented

**Reply by eifinger** [reply]
I enabled it via configuration.yaml for my local debug purposes.
```yaml
logger:
  default: info
  logs:
    homeassistant.components.waze_travel_time: debug
    homeassistant.components.google_travel_time: debug
    homeassistant.helpers.location: debug
````

**Reply by thecode** [reply]
I assume in it will be allowed to add `homeassistant.helpers.location` to the loggers list

**Comment by MartinHjelmare** [comment, marked as answer]
Debug logging can now be activated per integration.
